src/session/main.c

```diff
@@ -0,0 +1,11 @@
+/*
+ * D-Bus Session Initiator Main Entry
+ */
+
+#include <inttypes.h>
+
+extern int32_t dbrk_session_main(int32_t argc, uint8_t **argv);
```

Comment:
This isn't correct. `int32_t` is not guaranteed to be an alias of `int`, nor is `uint8_t` necessarily an alias of `unsigned char`.

- `int` is only required to be 16 bits wide, so the return value could be truncated on some weird MCU-like system. (To be fair, shell return values are truncated to unsigned 8 bits anyway, so it wouldn't make any difference.)
- `uint8_t` is an optional type and might not be defined at all. (Not a big deal, as it would just fail to compile in the first place.)
- `char` is only required to be _at least_ 8 bits wide, not exactly. In theory you could have a system with an 11-bit `char` or something, in which case a `char` and `uint8_t` could have different pointer representations.

I realize these are mostly theoretical concerns that won't matter on any supported platorm. You're also not actually redefining `main()` itself. Nevertheless I just think it's simply best to do bindings by the book. The increasingly popular use of exact-width types combined with incorrect assumptions about primitive types can turn out to be messy.

Anyway, I'm not a Rust guy myself, but based on a quick google search, I think you should probably use `libc::c_int` and `libc::c_char` here.

Reply:
This function refers to:

```rust
#[export_name = "dbrk_session_main"]
pub extern "C" fn main(
    argc: i32,
    argv: *const *const u8,
) -> i32 {
```

I think it matches the Rust function just fine. Please see the Rust-nomicon for details on FFI guarantees of Rust data-types, if unclear.

There is no `int`, or `char`, involved, nor any C datatypes that would ask for `libc::*` types. Am I missing something?